### PR TITLE
Check at compile time for CUDA-aware support

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp [hunk 1]
```diff
@@ -27,6 +27,16 @@
 
 #include <mpi.h>
 
+#if defined(ARBORX_USE_CUDA_AWARE_MPI)
+#include <mpi-ext.h>
```
Comment (Rombur) [inline src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp]: This doesn't exist for MPICH

**Comment by aprokop** [comment]
While at it, may want to look [here](https://www.olcf.ornl.gov/tutorials/gpudirect-mpich-enabled-cuda/).

**Comment by masterleinad** [comment]
With `spectrum-mpi` on `Summit` we get the warning which is fine with me.
On the other hand, we could also turn this check into one done at configure time.

**Comment by masterleinad** [comment]
> While at it, may want to look [here](https://www.olcf.ornl.gov/tutorials/gpudirect-mpich-enabled-cuda/).

I guess we could do more work to check if `CUDA-aware` MPI is supported, but at the moment I don't see the necessity to do it rigorously for all `MPI` implementations. I consider this an advanced feature that people should only use if they know what they are doing.

**Comment by Rombur** [comment]
@aprokop that was for Titan. Not sure if there is a machine we have access to where that works.

> With spectrum-mpi on Summit we get the warning which is fine with me.

Not with me. This is one of the platform that we **need** to support, we shouldn't throw a bunch of warnings when someone tries to use cuda-aware mpi.

> I consider this an advanced feature that people should only use if they know what they are doing.

I kind of disagree on that from the user perspective. A lot of people want to use cuda-aware MPI on Summit because it's supposed to be faster and it makes your code less complicated. So I think it's something a lot of people will want to turn on.

**Comment by sslattery** [comment]
I concur with @Rombur

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp [hunk 2]
```diff
@@ -27,6 +27,17 @@
 
 #include <mpi.h>
```
Comment (sslattery) [inline src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp]: Is there a way to do this at the CMake level instead? Or can we generate some kind of file like this that gets used at configure time to check for support?

Comment (masterleinad) [reply src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp]: Yes, we should be able to use `CHECK_CXX_SOURCE_COMPILE` here. This is what I was proposing in https://github.com/arborx/ArborX/pull/177#issuecomment-547615802.

**Comment by masterleinad** [comment]
> I kind of disagree on that from the user perspective. A lot of people want to use cuda-aware MPI on Summit because it's supposed to be faster and it makes your code less complicated. So I think it's something a lot of people will want to turn on.

Fine, but these people should know what they are doing and enable it judicisously. What I want to say is that I don't see a big problem if we warn about this once at compile or configure time.

**Comment by dalg24** [comment]
Don't get mixed up with CUDA-aware MPI in general and the algorithm proposed in this PR.  The changes proposed are not meant to be advertised to the user as is.  It is possible that we will enable selecting at runtime which algo to use.

**Comment by Rombur** [comment]
I am still against this PR. It doesn't work for any implementation that's based on mpich, like mvapich or cray, and it throws a warning on Summit.

**Pull request closed without merging**